src/kernels/gradient.cl

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2011-2018 Karlsruhe Institute of Technology
+ *
+ * This file is part of Ufo.
+ *
+ * This library is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+float compute_horizontal (global float *input,
+                          int idx,
+                          int idy,
+                          int width) {
+    int dx = 1;
+    if (idx == width - 1) {
+        dx = 0;
+    } else if (idx == 0) {
+        dx = 2;
+    }
+
+    return 0.5f * (input[idy * width + idx + dx] - input[idy * width + idx + dx - 2]);
```

Comment:
BTW this is how [numpy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.gradient.html) does it.